handler/src/main/java/io/netty/handler/ssl/OpenSslSessionCache.java

```diff
@@ -0,0 +1,493 @@
+                return false;
+            }
+
+            final NativeSslSession old = sessions.put(session.sessionId(), session);
+            if (old != null) {
+                notifyRemovalAndFree(old);
+            }
+        }
+        return true;
+    }
+
+    @Override
+    public final long getSession(long ssl, byte[] sessionId) {
+        OpenSslSessionId id = new OpenSslSessionId(sessionId);
+        final NativeSslSession session;
+        synchronized (this) {
+            session = sessions.get(id);
+            if (session == null) {
+                return -1;
+            }
+
+            // If the session is not valid anymore we should remove it from the cache and just signal back
+            // that we couldn't find a session that is re-usable.
+            if (!session.isValid() ||
+                    // This needs to happen in the synchronized block so we ensure we never destroy it before we
+                    // incremented the reference count. If we cant increment the reference count there is something
+                    // wrong. In this case just remove the session from the cache and signal back that we couldn't
+                    // find a session for re-use.
+                    !session.upRef()) {
+                // Remove the session from the cache. This will also take care of calling SSL_SESSION_free(...)
+                removeSessionWithId(session.sessionId());
+                return -1;
+            }
+
+            // At this point we already incremented the reference count via SSL_SESSION_up_ref(...).
+            if (session.shouldBeSingleUse()) {
+                // Should only be used once. In this case invalidate the session which will also ensure we remove it
+                // from the cache and call SSL_SESSION_free(...).
+                removeSessionWithId(session.sessionId());
+            }
+        }
+        session.updateLastAccessedTime();
+        return session.session();
+    }
+
+    void setSession(long ssl, String host, int port) {
+        // Do nothing by default as this needs special handling for the client side.
+    }
+
+    /**
+     * Remove the session with the given id from the cache
+     */
+    final synchronized void removeSessionWithId(OpenSslSessionId id) {
+        NativeSslSession sslSession = sessions.remove(id);
+        if (sslSession != null) {
+            notifyRemovalAndFree(sslSession);
+        }
+    }
+
+    /**
+     * Returns {@code true} if there is a session for the given id in the cache.
+     */
+    final synchronized boolean containsSessionWithId(OpenSslSessionId id) {
+        return sessions.containsKey(id);
+    }
+
+    private void notifyRemovalAndFree(NativeSslSession session) {
+        sessionRemoved(session);
+        session.free();
+    }
+
+    /**
+     * Return the {@link OpenSslSession} which is cached for the given id.
+     */
+    final OpenSslSession getSession(OpenSslSessionId id) {
+        synchronized (this) {
```

Comment:
The method could be synchronised, because its entire body is within a synchronised block.